Update remaining web area to use latest rgb() and hsl() syntax

FormularSumo opened the pull request:
Following on from https://github.com/mdn/content/pull/31453, this PR replaces all remaining rgba()/hsla() occurrences with rgb()/hsl() and updates all remaining rgb()/hsl() occurrences to use the latest syntax, within files\en-us\web

See https://github.com/mdn/content/issues/30844 for more details on why these changes are being made.

Included in these changes is:

- Reword api\canvas_api\tutorial\applying_styles_and_colors to use rgb instead of rgba and include a percentage value for alpha, and rename referenced canvas_rgba.png → canvas_rgb.png
- Remove reference to rgba in html\element\input\color

I have also left some instances of rgba/hsla in mathml\mathcolor, as it lists all the possible ways of using color, including rgb and rgba, new and old syntax, and in accessibility\understanding_color_and_luminance as it is referencing it as legacy syntax.

FormularSumo committed:
Update rgb/hsl syntax in web\accessibility

FormularSumo committed:
Update rgb & hsl to use new syntax

Including percentage instead of number for alpha value

FormularSumo committed:
Replace instances of rgba() with rgb() in web\api

Includes rewording of canvas_api\tutorial\applying)styles to use rgb instead of rgba and include a percentage value for alpha. Rename referenced canvas_rgba.png → canvas_rgb.png.

There weren't any instances of hsla()

FormularSumo committed:
Update rgb/hsl syntax in web\guide

FormularSumo committed:
Update rgb/hsl syntax in web\html

Remove reference to rgba in element\input\color

FormularSumo committed:
Update rgb/hsl syntax in web\mathml

I've left mathcolor as it lists all the possible ways of using color, including rgb and rgba, new and old syntax.

FormularSumo committed:
Update rgb/hsl syntax in web\mediia

FormularSumo committed:
Update rgb/hsl syntax in web\progressive_web_apps

FormularSumo committed:
Update rgb/hsl syntax in web\svg

FormularSumo committed:
Update miscellaneous rgb/hsl syntax in web

Including fixing a typo in css\color

FormularSumo committed:
Merge branch 'main' into rgba/hsla-syntax-updates

FormularSumo committed:
Apply linter/prettier suggestions

ShubhamOulkar commented on files/en-us/web/html/element/dialog/index.md:
```diff
@@ -338,7 +338,7 @@ dialog {
 /* Transition the :backdrop when the dialog modal is promoted to the top layer */
 dialog::backdrop {
-  background-color: rgb(0 0 0 / 0);
+  background-color: rgb(0 0 0 / 0%);
```
```suggestion
  background-color: rgb(0 0 0 / 0);
```

ShubhamOulkar commented on files/en-us/web/html/element/dialog/index.md:
```diff
@@ -348,15 +348,15 @@ dialog::backdrop {
 @starting-style {
   dialog[open]::backdrop {
-    background-color: rgb(0 0 0 / 0);
+    background-color: rgb(0 0 0 / 0%);
```
```suggestion
    background-color: rgb(0 0 0 / 0);
```

ShubhamOulkar commented on files/en-us/web/html/element/dialog/index.md:
```diff
@@ -462,11 +462,11 @@ dialog[open]::backdrop {
 @keyframes backdrop-fade-in {
   0% {
-    background-color: rgb(0 0 0 / 0);
+    background-color: rgb(0 0 0 / 0%);
```
```suggestion
    background-color: rgb(0 0 0 / 0);
```

ShubhamOulkar commented on files/en-us/web/html/element/input/color/index.md:
```diff
@@ -35,7 +35,7 @@ You can update the simple example above to set a default value, so that the colo
 {{EmbedLiveSample("Providing_a_default_color", 700, 30)}}
 
-If you don't specify a value, the default is `#000000`, which is black. The value must be in seven-character hexadecimal notation, meaning the "#" character followed by two digits each representing red, green, and blue, like this: `#rrggbb`. If you have colors that are in any other format (such as CSS color names or CSS color functions such as `rgb()` or `rgba()`), you'll have to convert them to  ...
+If you don't specify a value, the default is `#000000`, which is black. The value must be in seven-character hexadecimal notation, meaning the "#" character followed by two digits each representing red, green, and blue, like this: `#rrggbb`. If you have colors that are in any other format (such as CSS color names or CSS color functions such as `rgb()`), you'll have to convert them to hexadecimal  ...
```
```suggestion
If you don't specify a value, the default is `#000000`, which is black. The value must be in seven-character hexadecimal notation, meaning the "#" character followed by two digits each representing red, green, and blue, like this: `#rrggbb`. If you have colors that are in any other format (such as CSS color names or CSS color functions such as `rgb()` or `hsl()` ), you'll have to convert them to hexadecimal before setting the `value`.
```

ShubhamOulkar reviewed (commented):
there are some use cases, in which we must use commas in rgb() syntax. mostly to improve readability.
cc @dipikabh

FormularSumo committed:
Fix some incorrect syntax found in review

FormularSumo committed:
Update files/en-us/web/html/element/input/color/index.md

Co-authored-by: shubham oulkar <91728992+ShubhamOulkar@users.noreply.github.com>

FormularSumo committed:
Merge branch 'main' into rgba/hsla-syntax-updates

FormularSumo commented:
Hi @ShubhamOulkar, thanks for reviewing. I've fixed the errors you pointed out and committed the suggesiton to add ' or hsl()' to the color file. I see your point about having commas in some of the complex syntax to improve readability, I did wonder when making those changes and am happy to change back if needs be. I'm not sure about removing some of the percentage signs after 0 though. For consistency I think it looks better to have them, even if it doesn't actually make a different to its value. Especially as with the percentage sign is how it's been done in my previous pull request and the original fix(learn) area one.

ShubhamOulkar commented:
I already asked this questions to @dipikabh in another PR. Still waiting for a suggestion. I am going to add this PR reference there.

estelle commented on files/en-us/web/api/canvas_api/tutorial/applying_styles_and_colors/index.md:
```diff
@@ -158,27 +158,27 @@ draw();
 {{EmbedLiveSample("A_globalAlpha_example", "180", "180", "canvas_globalalpha.png")}}
 
-### An example using `rgba()`
+### An example using `rgb()`
```
should we say "An example using alpha transparency" instead, as the difference is not just the color type, but more so the change is color opacity?

FormularSumo replied:
Yeah that makes a lot of sense. I've gone with adding 'with alpha transparency' to the end of the sentence, as both rgb() and globalAlpha are ways of changing the alpha transparency, but this wording hopefully makes it clear how it's being done in the example.

estelle commented on files/en-us/web/api/mutationrecord/oldvalue/index.md:
````diff
@@ -37,7 +37,7 @@ The `logOldValue()` function is passed the `mutationRecords` array, which contai
 
 ```html
-<h1 id="h1" style="color:rgb(0, 0, 0);">Hi, Mom!</h1>
+<h1 id="h1" style="color:rgb(0 0 0);">Hi, Mom!</h1>
````
```suggestion
<h1 id="h1" style="color: rgb(0 0 0);">Hi, Mom!</h1>
```

estelle commented on files/en-us/web/api/popover_api/using/index.md:
```diff
@@ -498,11 +498,11 @@ html {
 
   100% {
-    background-color: rgb(0 0 0 / 0.25);
+    background-color: rgb(0 0 0 / 25%);
```
why are we changing valid float values to percentages?

ShubhamOulkar replied:
for consistency of documentation, we have decided after ` / ` we specify the % value for opacity.  But we can use numbers as well.

dipikabh replied:
Ah! @estelle, I've left comment here to say we are leaning towards `%` https://github.com/mdn/interactive-examples/pull/2703#issuecomment-1881475726 but decimal notation is not wrong. I believe Chris had a similar opinion on sticking with '%' in this [comment](https://github.com/mdn/content/pull/31453#pullrequestreview-1800279744).

estelle replied:
I see the "I like the idea of normalizing on % alpha values as a general rule." but I am unclear why. 
I hadn't seen your [comment](https://github.com/mdn/interactive-examples/pull/2703#issuecomment-1881475726). For reasons noted above -- that, unless something is officially a best practice from a standards body, we need to demonstrate there are different, equally valid ways of writing things. For MDN docs, which are referenced by teams writing best practices for their own engineering teams or frameworks, a sentence on the syntax/reference page does not suffice.

chrisdavidmills replied:
A few thoughts here to back up my statement:

1. Engineering teams use MDN to look up syntax, yes, but they also have coding standards/guidelines. MDN is no exception: https://developer.mozilla.org/en-US/docs/MDN/Writing_guidelines/Writing_style_guide/Code_style_guide. The CSS guidelines page doesn't mention using percentages in colors over numbers as a guideline, but I'd be in favor of adding it, for situations where teaching CSS color syntax isn't the main purpose of the example.
  - (1a) For situations where teaching CSS color syntax is the main purpose, of course, we should cover all the different syntax options in detail. That is essential. If this is not the case in the CSS Color ref pages and guides, then we should do work to address that.
  - (1b) If I approved any changes that removed different options from CSS Color ref pages and guides and replaced them all with %, then I'm sorry, and I should address that.
1. Having this as a guideline will make it easier for contributors to get started rather than something like "make sure you vary the syntax used across the color values in your examples".
2. It will also be confusing to try to enforce/review - how varied is varied enough? What values are acceptable? Which values are legacy and therefore invalid, versus valid values? Having one recommended syntax for general code examples takes all this confusion away.
3. It will also be harder to lint/auto-fix if we start adding such guidelines to the auto linter. imo, this could go hand in hand with not using legacy color syntax.

Ultimately, I don't think this issue needs a huge amount of time spent discussing it. We have much bigger fish to fry, and getting any such syntax normalized all over the site and kept up-to-date would start to become rather yak-shavey, unless we auto-linted /fixed it. But I do think guidelines are good, to help with consistency, and to help people not fall foul of using invalid syntax.

estelle replied:
Thanks @chrisdavidmills. That really clarifies the reasoning behind your comment. [My take](https://github.com/mdn/content/pull/31497#discussion_r1446416590) is the opposite, I concur that this isn't at all a hill to die on. Thanks

dipikabh replied:
Thanks all for sharing your viewpoints. To me, it was never about preferring one over the other but rather keeping the format similar. This is backed by content saying that both decimal and percentage are valid. And we don't mention any preference or recommendation of one over the other. When we're touching multiple repositories, especially in an exercise taking the help of the community, it is difficult to ensure that code in an example repository has the same variation as the code and examples in the corresponding content page. 

I agree with @estelle that it would be good for learners to see examples using both formats. And like @chrisdavidmills has mentioned, we could re-look at our color reference pages to make sure there are examples using both formats.

ShubhamOulkar commented on files/en-us/web/api/popover_api/using/index.md:
```diff
@@ -380,7 +380,7 @@ to take effect, as the specificity is the same */
 
 [popover]::backdrop {
-  background-color: rgb(0 0 0 / 0);
+  background-color: rgb(0 0 0 / 0%);
```
I got a reply for this you should write % sign with zeros, this is for consistency of documentation. But avoid large code projects.

FormularSumo committed:
Update files/en-us/web/api/mutationrecord/oldvalue/index.md

Co-authored-by: Estelle Weyl <estelle@openwebdocs.org>

FormularSumo committed:
Add 'with alpha transparency' to rgb() example

This is because this section is about using rgb() to specify alpha transparency instead of the globalAlpha property, so the focus should be on changing the transparency

FormularSumo commented on files/en-us/web/api/canvas_api/tutorial/applying_styles_and_colors/index.md:
```diff
@@ -43,9 +43,9 @@ function draw() {
+      ctx.fillStyle = `rgb(${Math.floor(255 - 42.5 * i)} ${Math.floor(
         255 - 42.5 * j,
-      )}, 0)`;
+      )} 0)`;
```
Unrelated to the rgb syntax, should there be a comma at the end of line 47? It was already there and not part of the rgb syntax so I left it but looking at it properly now it looks out of place, I wouldn't have thought it was needed at the end of that Math.floor argument, before the closing bracket?

dipikabh replied:
Good catch, @FormularSumo! The comma in `255 - 42.5 * j,` does seem out of place in Math.floor. I recommend opening a separate issue/PR for this so that the fix gets more eyes and is separate from these rgb changes.

FormularSumo replied:
I noticed the same thing with the comma on the draw function in line 74. I tried removing them and then running prettier on my system but in both cases it added the comma back in. Maybe it's something to do with it being the end of a line? I'm not too sure

dipikabh replied:
True, maybe. If you have the bandwidth and interest, you can take both these up in a separate PR.
Appreciate you paying attention to the details and testing locally 🙏

FormularSumo replied:
Cool, I might do that at some point, as well as the other weird formatting the linter is insisting on (with the rgb() over multiple lines)

dipikabh commented on files/en-us/web/api/canvasrenderingcontext2d/putimagedata/index.md:
```diff
@@ -99,9 +99,9 @@ function putImageData(
+      ctx.fillStyle = `rgb(${data[pos * 4 + 0]} ${data[pos * 4 + 1]} ${
         data[pos * 4 + 2]
-      }, ${data[pos * 4 + 3] / 255})`;
+      } / ${data[pos * 4 + 3] / 255})`;
```
same comment about two slashes and the last `255`

dipikabh commented on files/en-us/web/css/color/index.md:
```diff
@@ -37,7 +37,7 @@ color: rgba(34, 12, 64, 0.6);
 color: rgba(34 12 64 / 0.3);
 color: rgb(34 12 64 / 60%);
-color: rgba(34.6 12 64 / 30%);
+color: rgba(34 12 64 / 30%);
```
should this be:
```suggestion
color: rgb(34.6 12 64 / 30%);
```

FormularSumo replied:
- Updating the comment seems like a good idea, will do that now
- Link would useful I agree
- Some of them do have commas. The first 2 (`rgb` and `rbga`) are legaxy syntax - commas and floats. The other 4 are modern syntax, space-separate some with floats some with percentages, so I think it's fine as is. The other possible combination would be legacy comma-separated with percentages but that seems a bit redundant as it wouldn't ever really be used.

- Yeah that makes about the decimal, I've just looked at the `color_value` page you linked and it mentions floats being a later addition to rgb so it seems sensible to show it. It might be a good idea to add it to the last `rgb` and `hsl` example so they're in-line with the last `rgba` and `hsla` ones though?
- I suppose the difference between `rgb`/`hsl` (60%) and `hwb` (50%) opacity doesn't really matter. But within `rgb` I think it would make sense for all of them to be 60%, whereas currently the 4th and 6th are 30% instead. Having them consistent makes it more clear that `rgb` and `rgba` function the same, and brings it inline with the `hsl` and `hwb` examples which are consistent.

FormularSumo replied:
I have made those changes so you can see what I mean. What do you think of them?

I was wondering in terms of adding a link, would it make more sense to link to https://developer.mozilla.org/en-US/docs/Web/CSS/color_value given the [color page ](https://developer.mozilla.org/en-US/docs/Web/CSS/color) is about color generally and not just rgb?

dipikabh replied:
> Updating the comment seems like a good idea, will do that now

Wow, thanks!

> Some of them do have commas. The first 2 (rgb and rbga) are legaxy syntax - commas and floats. The other 4 are modern syntax, space-separate some with floats some with percentages, so I think it's fine as is. 

👍 

> Yeah that makes about the decimal, I've just looked at the color_value page you linked and it mentions floats being a later addition to rgb so it seems sensible to show it. It might be a good idea to add it to the last rgb and hsl example so they're in-line with the last rgba and hsla ones though?

Yeah, I see the updates. I think they look good and also stand out in the last examples for rgb/rgba and hsl/hsla. Thank you!

> I suppose the difference between rgb/hsl (60%) and hwb (50%) opacity doesn't really matter. But within rgb I think it would make sense for all of them to be 60%, whereas currently the 4th and 6th are 30% instead. Having them consistent makes it more clear that rgb and rgba function the same, and brings it inline with the hsl and hwb examples which are consistent.

I don't think they all should be the same necessarily within the rgb group. But maybe the pair could have been the same. For example, they could be:
```
color: rgb(34, 12, 64, 0.6);
color: rgba(34, 12, 64, 0.6);
color: rgb(34 12 64 / 0.4);
color: rgba(34 12 64 / 0.4);
color: rgb(34.6 12 64 / 50%);
color: rgba(34.6 12 64 / 50%);
```
I'm not sure if having the same opacity in all hsl/hsla examples is a good idea. Might be nice to show variation. But it doesn't hurt for them all to be `60%` at the moment.

FormularSumo replied:
I think for now I'll leave them all at 60% for now. Personally I prefer that, given the r g b values aren't varied I don't think the alpha should either, that makes it easier to just pick out the syntax differences to my eyes. But if sometimes else wants to change it later I don't really mind.

Just realised there is already a link to https://developer.mozilla.org/en-US/docs/Web/CSS/color_value right in the first sentence not sure how I missed that haha, so that's fine I reckon.

dipikabh commented on files/en-us/web/html/element/input/button/index.md:
```diff
@@ -233,7 +233,7 @@ const width = (canvas.width = window.innerWidth);
 const ctx = canvas.getContext("2d");
 
-ctx.fillStyle = "rgb(0,0,0)";
+ctx.fillStyle = "rgb(0,0,0%)";
```
Commas to be removed. No `%` needed here.

FormularSumo replied:
Whoops not sure what happened here, have fixed

dipikabh replied:
Thanks!

dipikabh commented on files/en-us/web/html/element/input/button/index.md:
```diff
@@ -272,7 +272,7 @@ canvas.onmouseup = () => {
 
 clearBtn.onclick = () => {
-  ctx.fillStyle = "rgb(0,0,0)";
+  ctx.fillStyle = "rgb(0,0,0%)";
```
Commas to be removed. No `%` needed here.

dipikabh reviewed (changes requested):
Thanks for these updates @FormularSumo 🙌 

I've added a few suggestions and questions.

FormularSumo commented on files/en-us/web/api/canvas_api/tutorial/applying_styles_and_colors/index.md:
```diff
@@ -158,27 +158,27 @@ draw();
   // Draw semi transparent rectangles
   for (let i = 0; i < 10; i++) {
-    ctx.fillStyle = `rgba(255, 255, 255, ${(i + 1) / 10})`;
+    ctx.fillStyle = `rgb(255 255 255 / ${(i + 1) / 10})`;
```
I don't think so, it being 10 means the alpha value is incremented from 0.1 to 1.0 as it loops, which make sense to show has transparency works. 10% would make it increment from 10 to 100, which would only work if it was a percentage.

FormularSumo committed:
Update files/en-us/web/api/canvas_api/tutorial/applying_styles_and_colors/index.md

Co-authored-by: Dipika Bhattacharya <dipika@foss-community.org>

FormularSumo committed:
Update files/en-us/web/api/canvas_api/tutorial/applying_styles_and_colors/index.md

Co-authored-by: Dipika Bhattacharya <dipika@foss-community.org>

FormularSumo committed:
Update files/en-us/web/api/canvas_api/tutorial/basic_animations/index.md

Co-authored-by: Dipika Bhattacharya <dipika@foss-community.org>

FormularSumo committed:
Fix incorrect rgb syntax in html area

FormularSumo commented on files/en-us/web/api/canvas_api/tutorial/pixel_manipulation_with_canvas/index.md:
```diff
@@ -106,7 +106,7 @@ function pick(event, destination) {
   const data = pixel.data;
 
-  const rgba = `rgba(${data[0]}, ${data[1]}, ${data[2]}, ${data[3] / 255})`;
+  const rgba = `rgb(${data[0]} ${data[1]} ${data[2]} / ${data[3] / 255})`;
```
I've just tested my change a local web server and it seems to be working fine, same as the original. When I console.log(data[3]) it returns 255 either way so I believe it does need still need to be divided by 255.

FormularSumo replied:
Renamed to rgbColor

dipikabh replied:
Nice, thanks for checking and confirming!

FormularSumo committed:
Rename rgba to rgbColor

FormularSumo committed:
Move related code onto the same line

FormularSumo commented on files/en-us/web/api/canvasrenderingcontext2d/putimagedata/index.md:
```diff
@@ -99,9 +99,9 @@ function putImageData(
       const pos = y * width + x;
-      ctx.fillStyle = `rgba(${data[pos * 4 + 0]}, ${data[pos * 4 + 1]}, ${
+      ctx.fillStyle = `rgb(${data[pos * 4 + 0]} ${data[pos * 4 + 1]} ${
         data[pos * 4 + 2]
```
I assume the linter check said it was too long at some point, so I've moved the ${ onto the next line like you've suggested and also moved the below line up as it seems like that should fit fine.

FormularSumo committed:
Move related code onto the same line

FormularSumo commented on files/en-us/web/css/css_images/using_css_gradients/index.md:
```diff
@@ -796,69 +796,62 @@ div {
+      rgb(143 77 63 / 25%) 10px
+    ), repeating-linear-gradient(45deg, transparent, transparent 5px, rgb(
+          143 77 63 / 25%
+        ) 5px, rgb(143 77 63 / 25%) 10px);
```
I think it needs an extra indent to what you've done, so I made a different commit to fix it with the indent. However prettier does seem to be complaining about it regardless

dipikabh replied:
@FormularSumo I think we need a comma here before between `5px` and `rgb` in:
`repeating-linear-gradient(45deg, transparent, transparent 5px
      rgb(143 77 63 / 25%) 5px,`

https://pr31497.content.dev.mdn.mozit.cloud/en-US/docs/Web/CSS/CSS_images/Using_CSS_gradients#plaid_gradient

FormularSumo replied:
Yes whoops I didn't mean to get rid of that, have just added back

FormularSumo replied:
Unfortunately the linter still doesn't like it. I've tried various different things locally running prettier and it's very stuborn that that is how it wants it to be formatted :/

dipikabh replied:
🤷‍♀️

FormularSumo committed:
Merge branch 'main' into rgba/hsla-syntax-updates

ShubhamOulkar commented on files/en-us/web/api/canvas_api/tutorial/pixel_manipulation_with_canvas/index.md:
```diff
@@ -106,7 +106,7 @@ function pick(event, destination) {
   const data = pixel.data;
 
-  const rgba = `rgba(${data[0]}, ${data[1]}, ${data[2]}, ${data[3] / 255})`;
+  const rgba = `rgb(${data[0]} ${data[1]} ${data[2]} / ${data[3] / 255})`;
```
I prefer legacy syntax inside template literals. It is less confusing.

dipikabh reviewed (changes requested):
@FormularSumo Thanks for your thorough work! We're getting very close, just a few last comments.

FormularSumo committed:
Update files/en-us/web/api/canvas_api/tutorial/applying_styles_and_colors/index.md

Co-authored-by: Dipika Bhattacharya <dipika@foss-community.org>

FormularSumo committed:
Add back comma in using_css_gradients

FormularSumo committed:
Update comments in color

FormularSumo committed:
Add back decimal in color, and add to rgb/hsl

I removed initially because it looked out of place, but actually it was probably to demonstrate having a decimal value (since it wasn't always supported in rgb). I've added it to rgb as well as rgba, and added an equivalent to hsl (in addition to hsla) to make it obvious that both versions work with decimals.

FormularSumo committed:
color syntax - make alpha values consistent

I don't think there was any reason to have different alpha values for these two rgba examples. The different syntaxes are to show how they're used, so changing the alpha value for just rgba seemed to imply that it somehow uses alpha differently from rgb.

This also brings it in line with the hsl/a examples which are consistent (also with 60% opacity)

fred-wang reviewed (approved):
> I have also left some instances of rgba/hsla in mathml\mathcolor, as it lists all the possible ways of using color,

I think that's fine if you remove them too. The corresponding "syntax" sections in https://developer.mozilla.org/en-US/docs/Web/CSS/color_value does not list them as an example either. The "Values" section of mathbackground/mathcolor also link to that `<color>` article where the formal syntax is detailed.

dipikabh commented:
@FormularSumo, thanks a lot for addressing all the comments!
I guess the only thing pending is the linter failing issue. Are you able to resolve it? We are ready to land this one after the lint test passes.

FormularSumo committed:
Remove legacy rgba/hsla syntax from mathml\mathcolor

I decided to make the decimal example .6 / .2 rather than .0 as I think that's better at showing using them, and is also consistent with https://developer.mozilla.org/en-US/docs/Web/CSS/color

FormularSumo committed:
Commit prettier suggestions

As strange as it appears to be

FormularSumo commented:
@dipikabh Yep the linter checks should have been resolved now. Unfortunately that means undoing the formatting change to using_css_gradients but not much that can be done in this PR at least about that. I've removed the legacy syntax from mathml\mathcolor like @fred-wang suggested. Should all be ready to be merged now!

FormularSumo committed:
Merge branch 'main' into rgba/hsla-syntax-updates

dipikabh reviewed (approved):
Awesome work, @FormularSumo! Appreciate your efforts and diligence in updating the rgb/hsl syntax to the latest version. Thanks for your contribution to keep the MDN docs updated.
